src/middlewared/middlewared/plugins/zfs_/events_linux.py

```diff
@@ -0,0 +1,59 @@
+import os
+import libzfs
+import multiprocessing
+import time
+
+from middlewared.utils import start_daemon_thread
+from middlewared.utils.osc import set_thread_name
+
+
+def zfs_events(child_conn, start_time):
+    zevent_fd = None
+    try:
+        zevent_fd = os.open('/dev/zfs', os.O_RDWR)
+        with libzfs.ZFS() as zfs:
+            while True:
+                event = zfs.zpool_events_single(zevent_fd)
+                if event.get('time'):
+                    # When we retrieve zfs events, we start from a point where we retrieve old events as well,
+                    # so in this case we filter the old events out by comparing the timestamp of the event
+                    # with the time events endpoint was called and only sending events for those zfs events
+                    # which came after that timestamp
+                    if start_time > float(f'{event["time"][0]}.{event["time"][1]}'):
+                        continue
+
+                child_conn.send(event)
+    finally:
+        if zevent_fd is not None:
+            os.close(zevent_fd)
+
+
+def setup_zfs_events_process(middleware):
+    set_thread_name('retrieve_zfs_events_thread')
+    while True:
+        start_time = time.time_ns() / (10 ** 9)
```

Comment:
This is non-monotonic time which can cause issues when system clock is adjusted. We need other way to solve this issue. What "old" events are coming in?

Reply:
So basically when we start reading events with opening a fd to /dev/zfs, we start from the first event zfs has logged and then complete the trail of events coming to a blocking call where we wait for new events to occur. This means that if a scrub was run before, we would have old events of scrub if the thread starts after that scrub and we need to skip those old events and only consider new events given to us.